### DISCUSSION TITLE
Connecting to an IP address

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#2903] [Q&A] [answered]
I'm trying to connect to an IP address, using `edgedb --dsn edgedb://127.0.0.1:10700` but this returns an error:

> edgedb error: Cannot use `verify_hostname` or system root certificates with an IP address

* Adding the ` --no-tls-verify-hostname` option doesn't help either.
* using 1-beta-3
* 127.0.0.1 is only used for testing, I actually want to use the IP of a VM

**Comment by fantix** [comment, marked as answer]
According to https://github.com/edgedb/edgedb-rust/pull/68/commits/fb38c7577996fd5ef515afab444fcedae5cd6a4a:

An [upstream issue](https://github.com/rustls/rustls/issues/184) is temporarily stopping the EdgeDB CLI from using IPs as server hosts, and the current fix (@tailhook please correct me) is to transparently use `127.0.0.1.host-for-ip.edgedb.net` with `--no-tls-verify-hostname`. I think we don't have `host-for-ip.edgedb.net` ready yet (@elprans ?). Once that's ready, you could use `edgedb --dsn edgedb://127.0.0.1:10700 --no-tls-verify-hostname` to connect to an IP address. However, the recommended way to connect to a remote EdgeDB instance is to use `edgedb instance link` so that you don't have to deal with the certificates manually.

For the moment, you could either use non-IP host names like `localhost` or something in `/etc/hosts`, or use services like [nip.io](https://github.com/exentriquesolutions/nip.io) temporarily (while `xip.io` is down).

**Reply by CodesInChaos** [reply]
`instance link` failed with the same error when using an IP (I already tried that), but adding an entry to the hosts file helped.

Manually substituting the IP address as host in the credentials file also seems to work, so the hostname is no longer required after the initial setup.

I guess it'd also have been possible to copy the credentials file from the VM the edgedb server is running on, adding a `"host":"<ip>"` key manually.

**Comment by tailhook** [comment]
Hm, yes, you have to specify both `verify_hostname: false` (which is default for self-signed certificate) and certificate (to avoid use of system certificates). So you can fetch public certificate somehow and do this:
```
edgedb -H 127.0.0.1 -P 5656 --tls-ca-file cert.txt
edgedb instance link -H 127.0.0.1 -P 5656 --tls-ca-file cert.txt
```

This also explains why changing host in JSON works.

To make it easier we have to:
1. Allow specify `edgedb://...?tls_ca_file=` (currently you can't specify certificate for DSN)
2. Ensure `--trust-tls-cert` works when using `link` (probably by using host hack @fantix mentioned above, but that's implementation detail)
